### ISSUE TITLE
Support for forced inline?

### ISSUE THREAD

**Issue by kirbyfan64** [nim-lang/Nim#2751]
Some compilers (including GCC, Clang, and MSVC) support a way to "force" inlining. Could this be added to Nim in the form of a `forceInline` pragma? I'd be happy to add it if it's wanted.

**Comment by kirbyfan64** [comment]
Like I said, I'd be happy to take a shot at it...

**Comment by Araq** [comment]
No, it's not wanted. Instead map Nim's inline to force inline for the C compilers, preferably after  you demonstrated it's necessary.

**Comment by kirbyfan64** [comment]
@Araq Well, an answer is better than no answer. :)

I just know that (mainly in the C++ world) compilers (mainly GCC) can be stubborn about actually inlining stuff. I'll toy with a concrete use case when I get access to a computer.

**Comment by def-** [comment]
Nim's templates can be used to force inlining.

**Comment by kirbyfan64** [comment]
But wasn't there a discussion at one point about trying to replace a template in a PR with an inline proc because it was cleaner?

**Comment by josephwecker** [comment]
Compilers have the ability to force inline vs hinting because there is a difference between

* [force] "I need this inlined for ABI compatibility- non-negotiable" (would also cause error if recursion or something makes inlining impossible) versus
* [hint] "Performance requires this to be inlined even if the user is compiling for size instead of speed" versus
* [unnecessary hint, usual usage] "I'm trying to do the compiler's job and decide what should be inlined."

In my opinion Nim should have a philosophy that certain pragmas are **(1) semantic: necessary** to implement certain kinds of compatibility (asmNoStackFrame, global, ...) versus pragmas that are **(2) efficiency hints** (unroll, linearScanEnd, acyclic, ...) that do not affect the semantics of the program and could be considered workarounds for deficiencies in what is ostensibly the compiler's responsibility. The former should be rare because those compatibility needs are rare. The latter should be rare because it indicates a "bug" in the compiler's lower level optimization (that may or may not have a mathematically tractable solution right now).

The thing about inlining (and possibly dead code elimination) is that they exist in _both_ categories for different reasons. "Force inlining and error if you can't" for ABI compatibility is a separate language feature from "Please help my dumb backend compiler realize that this should probably be inlined- but if its a smart compiler disregard this," and in my opinion they should have two separate pragmas, at least until the latter (non-force) can be deprecated because we're sure all compilers "do the optimal thing."

**Comment by ozra** [comment]
Mind you, the _always_inline_ hints inline even in debug-mode as far as I know, which doesn't seem like a common intention.

I depend a lot on inlining in a lot of code I code (well, I'm talking C++ now), and find that you have quite good control of inlining behavior via gcc compiler flags (clang is a dev-compiler in my eyes, don't know about inlining-perf there..). So: do you have a specific culprit in mind that would need it?

**Comment by kirbyfan64** [comment]
1. Clang can actually be more standards-conformant than GCC at times!

2. Not really. Maybe it's my C++ experience knocking on my head to force inline stuff...

I need to find an example where forced inlining makes a differenc.

**Comment by ozra** [comment]
@kirbyfan64 
1. Agreed, I was referring more to that clang is superior in error messages and compiles faster, thus I use it when dev'ing. But so far gcc has always made faster code for me, so I used that for release compilation.
2. I might very well have missed some inlinings without knowing, so I've perhaps talked a bit out my ass in the above. I don't analyze asm output of all my code ;)

Also, I just came to think about an actual use case for forced inlining: One has a function which one splits up in several smaller ones for readability. They are used _only_ from this "main" function. They are too big for the compiler to consider as inline candidates. Baam. A force-inline pragma would be important here, and OK even in debug mode. 

I retract my earlier statements.

This would actually be beneficial to be able to state from Nim. Forcing all inlines could be counter productive (let backend take _most_ decisions - it's _usually_ better).
